aten/src/ATen/native/Bincount.cpp

```diff
@@ -0,0 +1,77 @@
+// Returns the frequency of elements of input non-negative integer tensor.
+
+#include "ATen/ATen.h"
+#include "ATen/Dispatch.h"
+
+#include <tuple>
+
+namespace at { namespace native {
+
+#define AT_DISPATCH_BINCOUNT_INT_CASE(scalar, type)                \
+  case scalar: {                                                   \
+    return _bincount_cpu_template<type>(self, weights, minlength); \
+  }
+
+namespace {
+template <typename integral_t>
+Tensor _bincount_cpu_template(
+    const Tensor& self,
+    const Tensor& weights = {},
+    int64_t minlength = 0) {
+  if (minlength < 0) {
+    throw std::domain_error("minlength should be >= 0");
+  }
+  if (self.dim() != 1 || !isIntegralType(self.type().scalarType()) ||
+      *self.min().data<integral_t>() < 0) {
+    throw std::domain_error(
+        "bincount only supports 1-d non-negative integral inputs.");
+  }
+
+  bool isWeights = weights.defined();
+  if (isWeights && weights.numel() != self.numel()) {
+    throw std::runtime_error("input and weights should have the same length");
+  }
+
+  Tensor output;
+  int64_t nbins = static_cast<int64_t>(*self.max().data<integral_t>()) + 1L;
+  nbins = std::max(nbins, minlength); // at least minlength # of bins
+
+  integral_t* self_p = self.contiguous().data<integral_t>();
+  if (isWeights) {
+    output = zeros(CPU(kDouble), {nbins});
```

Comment:
I think this one should match the weights. It will be quite important in the CUDA case, because double precision math is quite slow

Reply:
Also cc @gchanan who investigated type promotion for various functions